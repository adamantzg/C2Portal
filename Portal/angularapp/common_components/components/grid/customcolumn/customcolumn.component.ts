import { Component, OnInit, Input, Output, ComponentFactoryResolver,
  ViewContainerRef, Injector, ViewChild } from '@angular/core';
import { GridColumn } from '../grid.component';

@Component({
  selector: 'app-customcolumn',
  templateUrl: './customcolumn.component.html',
  styleUrls: ['./customcolumn.component.css']
})
export class CustomcolumnComponent implements OnInit {

  constructor(private resolver: ComponentFactoryResolver,
    private injector: Injector,
    private vcr: ViewContainerRef ) { }

  @Input()
  row: any;
  @Input()
  column: GridColumn;
  component: any;

  @ViewChild('content', { read: ViewContainerRef }) content;

  ngOnInit() {
    const factory = this.resolver.resolveComponentFactory(this.column.customComponentType);
    this.component = this.content.createComponent(factory);
    this.component.instance.row = this.row;
    this.component.instance.column = this.column;
  }

}

export interface ICustomGridColumnComponentContent {
  row: any;
  column: GridColumn;
}
