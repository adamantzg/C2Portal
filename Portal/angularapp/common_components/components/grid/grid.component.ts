import { Component, OnInit, Input, Output, EventEmitter, ComponentFactoryResolver, ViewContainerRef, Type } from '@angular/core';
import { DatePipe, DecimalPipe, CurrencyPipe, PercentPipe } from '@angular/common';
import { FormGroup, FormControl } from '@angular/forms';
import { TypeaheadMatch } from 'ngx-bootstrap';

@Component({
  selector: 'app-grid',
  templateUrl: './grid.component.html',
  styleUrls: ['./grid.component.css'],
  providers: [DatePipe, DecimalPipe, CurrencyPipe, PercentPipe]
})
export class GridComponent implements OnInit {

  constructor(private componentFactoryResolver: ComponentFactoryResolver,
    private viewContainerRef: ViewContainerRef) { }

  private _definition: GridDefinition;

  @Input()
  data: any[];
  @Input()
  get definition(): GridDefinition {
    return this._definition;
  }
  set definition(value: GridDefinition) {
    this._definition = value;
    this.onDefinitionSet(value);
  }
  @Output()
  ButtonClicked = new EventEmitter();
  columnTypes = Object.assign({}, GridColumnType);
  dataTypes = Object.assign({}, ColumnDataType);
  fixedHeaderStyle: any;
  @Input()
  style = {};
  @Input()
  filter: string;
  @Output()
  SortChange = new EventEmitter();
  @Input()
  css = 'table';
  @Input()
  keyField = 'id';
  @Input()
  editMode = GridEditMode.NoEdit;
  @Input()
  editedRow: any;
  editModes = Object.assign({}, GridEditMode);
  editorTypes = Object.assign({}, GridEditorType);

  editValidation = false;

  form: FormGroup;

  ngOnInit() {
    this.fixedHeaderStyle = { height: this.calculateScrollableHeight(this.style['height']), 'overflow-y': 'auto'};
  }

  buttonClicked(d: any, c: GridColumn, button?: GridButton) {
    if (button != null) {
      if (!button.requiresValidation || this.form.valid) {
        this.ButtonClicked.emit(new GridButtonEventData(c.name, d, button.name));
      }
      if (button.requiresValidation && this.form.invalid) {
        this.editValidation = true;
      }
    } else {
      this.ButtonClicked.emit(new GridButtonEventData(c.name, d));
    }

  }

  calculateScrollableHeight(styleHeight: string) {
    if ( styleHeight != null) {
      const num = +styleHeight.replace('px', '');
      if (num > 0) {
        return (num - 40).toString() + 'px';
      }
    }
    return '';
  }

  getValue(row: any, c: GridColumn) {

    if (c.fieldValueCallback == null) {
      let obj = row;
      const parts = c.field.split('.');
      for (let i = 0; i < parts.length - 1; i++) {
        obj = obj[parts[i]];
      }
      return obj[parts[parts.length - 1]];
    }
    return c.fieldValueCallback(row);
  }

  checkSort(c: GridColumn) {
    return this.definition.sort != null && this.definition.sort.column.field === c.field;
  }

  toggleSort(c: GridColumn) {
    if (this.definition.sort == null) {
      this.definition.sort = {column : c, direction: SortDirection.Asc};
    } else {
      let direction = SortDirection.Asc;
      if (this.definition.sort.column.field === c.field && this.definition.sort.direction === SortDirection.Asc) {
        direction = SortDirection.Desc;
      }
      this.definition.sort.column = c;
      this.definition.sort.direction = direction;

    }
    this.SortChange.emit(this.definition.sort);
  }

  getSortIconClass(c: GridColumn) {
    return `glyphicon glyphicon-arrow-${this.definition.sort.direction === SortDirection.Asc ? 'up' : 'down' } sort-icon`;
  }

  checkColumnVisibility(c: GridColumn) {
    if (this.definition.columnVisibilityCallback == null) {
      return true;
    }
    return this.definition.columnVisibilityCallback(c);
  }

  checkColumnButtonVisibility(row: any, c: GridColumn, button: string) {
    if (this.definition.columnButtonVisibilityCallback == null) {
      return true;
    }
    return this.definition.columnButtonVisibilityCallback(row, c.name, button);
  }

  checkColumnGroupDisabledStatus(row: any, c: GridColumn, button: string) {
    if (this.definition.columnButtonDisabledStatusCallback == null) {
      return false;
    }
    return this.definition.columnButtonDisabledStatusCallback(row, c.name, button);
  }

  checkColumnEditorVisibilityStatus(row: any, c: GridColumn) {
    if (this.definition.columnEditorVisibilityStatusCallback == null) {
      return true;
    }
    return this.definition.columnEditorVisibilityStatusCallback(row, c);
  }

  toggleValue(c: GridColumn) {
    this.data.forEach(d => d[c.field] = c.selected);
  }

  checkMode(c: GridColumn) {
    return c.editModes.find(m => m === this.editMode) != null;
  }

  onDefinitionSet(def: GridDefinition) {
    const controls = {};
    def.columns.forEach(c => {
      if (c.editable) {
        controls[c.name] = new FormControl(c.name);
      }
    });
    this.form = new FormGroup(controls);
  }

  onAutoCompleteSelected(c: GridColumn, $event: TypeaheadMatch ) {
    if (c.autoCompleteData.onSelectedCallback != null) {
      c.autoCompleteData.onSelectedCallback($event.item);
    }
  }

  onAutoCompleteLoading(c: GridColumn, $event: boolean) {
    if (c.autoCompleteData.loadingCallback != null) {
      c.autoCompleteData.loadingCallback($event);
    }
  }

}

export class GridDefinition {
  columns: GridColumn[];
  fixedHeaders = false;
  multiSelect = false;
  scrollHeight: string;
  sort: Sort;
  columnVisibilityCallback: (c: GridColumn) => boolean;
  columnButtonVisibilityCallback: (row: any, name: string, button: string) => boolean;
  columnButtonDisabledStatusCallback: (row: any, name: string, button: string) => boolean;
  columnEditorVisibilityStatusCallback: (row: any, c: GridColumn) => boolean;

  constructor(columns: GridColumn[], fixedHeaders?: boolean, multiSelect?: boolean, scrollHeight?: string) {
    this.columns = columns;
    if (fixedHeaders != null) {
      this.fixedHeaders = fixedHeaders;
    }
    if (multiSelect != null) {
      this.multiSelect = multiSelect;
    }
    if (scrollHeight != null) {
      this.scrollHeight = scrollHeight;
    }
  }

  getColumn(name: string) {
    return this.columns.find(c => c.name === name);
  }
}

export class GridColumn {
  title: string;
  field: string;
  name: string;
  style: any;
  buttonCss: string;
  customComponentType: Type<any>;
  type: GridColumnType;
  data: any;
  sortable: boolean;
  dataType: ColumnDataType;
  format: any;
  selected: boolean; // used for checbox all selection
  buttonIcon: string;
  editorType: GridEditorType;
  editable = true;
  buttons: GridButton[] = [];
  editModes: GridEditMode[] = [GridEditMode.NoEdit];
  required: boolean;
  valueField: string; // dropdown data id
  displayField: string; // dropdown data title
  selectedValueField: string; // field in data to be updated by valueField
  dropdownData: any[];
  autoCompleteData: ColumnAutoCompleteData;
  multipleSelectionData: MultipleSelectionData;
  fieldValueCallback: any;

  constructor(title: string, field: string, type?: GridColumnType, name?: string, style?: any,
    buttonCss?: string, customComponentType?: Type<any>, data?: any, sortable?: boolean,
    dataType?: ColumnDataType, format?: any, buttonIcon?: string, editable?: boolean,
    editorType?: GridEditorType, buttons?: GridButton[], editModes?: GridEditMode[],
    required?: boolean, autoCompleteData?: ColumnAutoCompleteData, multipleSelectionData?: MultipleSelectionData,
     fieldValueCallback?: any) {
    this.title = title;
    this.field = field;
    this.name = name;
    if (type != null) {
      this.type = type;
    } else {
      this.type = GridColumnType.Label;
    }
    this.style = style;
    this.buttonCss = buttonCss;
    if (name == null) {
      this.name = this.title;
    }
    this.customComponentType = customComponentType;
    this.data = data;
    this.sortable = sortable;
    this.dataType = dataType;
    if (this.dataType == null) {
      this.dataType = ColumnDataType.Text;
    }
    this.format = format;
    this.buttonIcon = buttonIcon;
    this.editorType = editorType;
    if (editable != null) {
      this.editable = editable;
    }

    if (this.editorType == null) {
      this.editorType = GridEditorType.Textbox;
    }
    this.buttons = buttons;
    if (editModes != null) {
      this.editModes = editModes;
    }
    this.required = required;
    this.autoCompleteData = autoCompleteData;
    this.multipleSelectionData = multipleSelectionData;
    this.fieldValueCallback = fieldValueCallback;
  }

}

export class GridButton {
  css: string;
  icon: string;
  text: string;
  name: string;
  requiresValidation = false;

  constructor(text: string, name: string, icon?: string, css?: string, requiresValidation?: boolean) {
    this.text = text;
    this.name = name;
    this.css = css;
    this.icon = icon;
    if (requiresValidation) {
      this.requiresValidation = requiresValidation;
    }
  }
}

export class ButtonGroupColumn extends GridColumn {
  buttons: GridButton[] = [];
  editModes: GridEditMode[] = [GridEditMode.NoEdit];

  constructor(title: string, buttons: GridButton[], editModes?: GridEditMode[]) {
    super(title, '', GridColumnType.ButtonGroup);
    this.buttons = buttons;
    if (editModes != null) {
      this.editModes = editModes;
    }
  }
}

export class GridButtonEventData {
  column: string;
  row: any;
  name: string;

  constructor(column: string, row: any, name?: string) {
    this.column = column;
    this.row = row;
    this.name = name;
  }
}

export class CustomColumnEventData {
  name: string;
  row: any;
}

export enum GridColumnType {
  Label = 0,
  Checkbox = 1,
  Button = 2,
  Custom = 3,
  Checkmark = 4,
  ButtonGroup = 5
}

export enum ColumnDataType {
  Text,
  Numeric,
  Date,
  Currency,
  Percent
}

export enum SortDirection {
  Asc = 0,
  Desc = 1
}

export class Sort {
  column: GridColumn;
  direction: SortDirection;

  constructor(column: GridColumn, direction: SortDirection) {
    this.column = column;
    this.direction = direction;
  }
}

export enum GridEditMode {
  NoEdit,
  AddNew,
  Edit
}

export enum GridEditorType {
  Textbox,
  Checkbox,
  Dropdown,
  Autocomplete,
  MultipleChoiceTagInput,
  MultipleChoiceCheckboxes
}

export class ColumnAutoCompleteData {
  data: any[];
  onSelectedCallback: (obj: any) => any;
  optionsLimit = 20;
  loadingCallback: (isLoading: boolean) => any;
  minLength = 3;
  optionField: string;

  constructor(data: any[], optionsLimit: number, minLength: number, optionField: string, onSelectedCallback: any, loadingCallback: any) {
    this.data = data;
    this.optionsLimit = optionsLimit;
    this.minLength = minLength;
    this.optionField = optionField;
    this.onSelectedCallback = onSelectedCallback;
    this.loadingCallback = loadingCallback;
  }
}

export class MultipleSelectionData {
  autocompleteItems: any[];
  autocompleteOnly = false;
  idField = 'id';
  displayField = 'name';
  placeHolder = 'Choose...';
  selectedField = 'selected';

  constructor(autocompleteOnly?: boolean, autocompleteItems?: any[], idField: string = 'id',
    displayField: string = 'name', selectedField: string = 'selected', placeHolder?: string ) {
    this.autocompleteItems = autocompleteItems;
    this.autocompleteOnly = autocompleteOnly;
    this.idField = idField;
    this.displayField = displayField;
    this.selectedField = selectedField;
    this.placeHolder = placeHolder;
  }
}




