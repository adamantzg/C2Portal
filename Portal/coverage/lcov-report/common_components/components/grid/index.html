<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for common_components/components/grid</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="..\..\..\prettify.css" />
    <link rel="stylesheet" href="..\..\..\base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(..\..\..\sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="..\..\..\index.html">All files</a> common_components/components/grid
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">83.01% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>215/259</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">58.33% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>70/120</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">78% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>39/50</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">82.77% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>197/238</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line high'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file medium" data-value="formatPipe.ts"><a href="formatPipe.ts.html">formatPipe.ts</a></td>
	<td data-value="76.19" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 76%;"></div><div class="cover-empty" style="width:24%;"></div></div></td>
	<td data-value="76.19" class="pct medium">76.19%</td>
	<td data-value="21" class="abs medium">16/21</td>
	<td data-value="53.13" class="pct medium">53.13%</td>
	<td data-value="32" class="abs medium">17/32</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="70.59" class="pct medium">70.59%</td>
	<td data-value="17" class="abs medium">12/17</td>
	</tr>

<tr>
	<td class="file high" data-value="grid.component.html"><a href="grid.component.html.html">grid.component.html</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	</tr>

<tr>
	<td class="file high" data-value="grid.component.ts"><a href="grid.component.ts.html">grid.component.ts</a></td>
	<td data-value="83.54" class="pic high"><div class="chart"><div class="cover-fill" style="width: 83%;"></div><div class="cover-empty" style="width:17%;"></div></div></td>
	<td data-value="83.54" class="pct high">83.54%</td>
	<td data-value="237" class="abs high">198/237</td>
	<td data-value="60.23" class="pct medium">60.23%</td>
	<td data-value="88" class="abs medium">53/88</td>
	<td data-value="76.6" class="pct medium">76.6%</td>
	<td data-value="47" class="abs medium">36/47</td>
	<td data-value="83.64" class="pct high">83.64%</td>
	<td data-value="220" class="abs high">184/220</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Wed Sep 05 2018 16:05:12 GMT+0200 (Central European Summer Time)
</div>
</div>
<script src="..\..\..\prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="..\..\..\sorter.js"></script>
<script src="..\..\..\block-navigation.js"></script>
</body>
</html>
