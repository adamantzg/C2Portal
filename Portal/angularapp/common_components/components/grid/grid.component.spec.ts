import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { GridComponent, GridDefinition, GridColumn, GridEditMode, GridColumnType, GridButton, GridButtonEventData } from './grid.component';
import { FormsModule, ReactiveFormsModule, FormGroup } from '@angular/forms';
import { DatePipe, DecimalPipe, CurrencyPipe } from '@angular/common';
import { ColumnFormatPipe } from './formatPipe';
import { CustomcolumnComponent } from './customcolumn/customcolumn.component';
import { TypeaheadModule } from 'ngx-bootstrap';
import * as $ from 'jquery';
import { CommonAppModule } from '../../common.module';
import { TagInputModule } from 'ngx-chips';

describe('GridComponent', () => {
  let component: GridComponent;
  let fixture: ComponentFixture<GridComponent>;

  setupTestBed({
      providers: [DatePipe, DecimalPipe, CurrencyPipe],
      declarations: [],
      imports: [FormsModule, ReactiveFormsModule, TypeaheadModule, CommonAppModule, TagInputModule]
    });

  beforeEach(() => {
    fixture = TestBed.createComponent(GridComponent);
    component = fixture.componentInstance;

  });

  test('should create', () => {
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  test('hidden editors', () => {
    const def = new GridDefinition([]);
    let c = new GridColumn('a', 'a', GridColumnType.Label, 'a');
    c.editable = true;
    def.columns.push(c);
    c = new GridColumn('b', 'b', GridColumnType.Label, 'b');
    c.editable = true;
    def.columns.push(c);
    const data = [{id: 1, a: 1, b: 2}, {id: 2, a: 3, b: 4}];
    component.data = data;
    component.editMode = GridEditMode.Edit;

    component.definition = def;

    component.editedRow = JSON.parse(JSON.stringify(data[0]));
    component.definition.columnEditorVisibilityStatusCallback = (row, col) =>  row.id === 1 && col.name === 'a' ? false : true;

    fixture.detectChanges();

    expect($('input').length).toBe(1);
    expect($('input[name=a]').length).toBe(0);

    component.editedRow = JSON.parse(JSON.stringify(data[1]));
    fixture.detectChanges();
    expect($('input').length).toBe(2);

  });

  test('buttonClicked', () => {
    component.form = new FormGroup({});
    const b = new GridButton('a', 'a');
    const c = new GridColumn('c', 'c', null, 'c');
    let event: GridButtonEventData;
    component.ButtonClicked.subscribe((e) => event = e );
    component.buttonClicked({}, c, b );
    expect(event.column).toBe(c.name);
    expect(event.name).toBe(b.name);
    expect(event.row).toBeTruthy();

    b.requiresValidation = true;
    jest.spyOn(component.form, 'valid', 'get').mockImplementationOnce(() => true);
    event.name = '';
    component.buttonClicked({}, c, b );
    expect(event.name).toBe(b.name);

    jest.spyOn(component.form, 'valid', 'get').mockImplementationOnce(() => false);
    jest.spyOn(component.form, 'invalid', 'get').mockImplementationOnce(() => true);
    component.buttonClicked({}, c, b );
    expect(component.editValidation).toBeTruthy();

  });

  test('calculateScrollableHeight', () => {
    let sh = null;
    expect(component.calculateScrollableHeight(sh)).toEqual('');

    sh = '100px';
    expect(component.calculateScrollableHeight(sh)).toEqual('60px');

    sh = 'a';
    expect(component.calculateScrollableHeight(sh)).toEqual('');

    sh = '20px';
    expect(component.calculateScrollableHeight(sh)).toEqual('-20px');
  });

  test('getValue', () => {

  });
});
