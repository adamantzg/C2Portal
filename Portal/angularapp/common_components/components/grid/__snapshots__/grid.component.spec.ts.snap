// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridComponent should create 1`] = `
<app-grid
  ButtonClicked={[Function EventEmitter]}
  SortChange={[Function EventEmitter]}
  columnTypes={[Function Object]}
  componentFactoryResolver={[Function CodegenComponentFactoryResolver]}
  css={[Function String]}
  dataTypes={[Function Object]}
  editMode="0"
  editModes={[Function Object]}
  editValidation="false"
  editorTypes={[Function Object]}
  keyField={[Function String]}
  style={[Function Object]}
  viewContainerRef={[Function ViewContainerRef_]}
>
  <form
    novalidate=""
  >
    <div>
      <table
        style="margin-bottom: 0px"
      >
        <thead>
          <tr>
            
          </tr>
        </thead>
        
      </table>
      
      
      
    </div>
  </form>
</app-grid>
`;
